File: app/routers/results.py
from beanie import PydanticObjectId
from fastapi import APIRouter, Depends, status

from app.core.dependencies import get_current_user, require_admin, require_user
from app.models.user import User
from app.schemas.common import DataResponse, ListResponse
from app.schemas.result import (
    CorrectionResponse,
    MatchResultCorrection,
    MatchResultCreate,
    MatchResultResponse,
    MatchResultUpdate,
    ResultRevisionResponse,
    TournamentResultResponse,
)
from app.schemas.scoring_rule import ScoringRuleResponse, ScoringRuleUpsert
from app.services import result_service

router = APIRouter(prefix="/api/tournaments", tags=["results"])
admin_matches_router = APIRouter(
    prefix="/api/admin/matches", tags=["admin:results"], dependencies=[Depends(require_admin)]
)
admin_results_router = APIRouter(
    prefix="/api/admin/results", tags=["admin:results"], dependencies=[Depends(require_admin)]
)
admin_tournament_router = APIRouter(
    prefix="/api/admin/tournaments", tags=["admin:results"], dependencies=[Depends(require_admin)]
)


@router.get(
    "/{tournament_id}/results",
    response_model=ListResponse[TournamentResultResponse],
    summary="List published (calculated) results for a tournament",
    dependencies=[Depends(require_user)],
)
async def list_results(tournament_id: PydanticObjectId) -> ListResponse[TournamentResultResponse]:
    results = await result_service.list_published_results(tournament_id)
    return ListResponse(data=[TournamentResultResponse.model_validate(r, from_attributes=True) for r in results])


@router.get(
    "/{tournament_id}/scoring-rule",
    response_model=DataResponse[ScoringRuleResponse],
    summary="Get the active scoring rule for a tournament",
    dependencies=[Depends(require_user)],
)
async def get_scoring_rule(tournament_id: PydanticObjectId) -> DataResponse[ScoringRuleResponse]:
    rule = await result_service.get_active_scoring_rule(tournament_id)
    return DataResponse(data=ScoringRuleResponse.model_validate(rule, from_attributes=True))


@admin_tournament_router.post(
    "/{tournament_id}/scoring-rule",
    response_model=DataResponse[ScoringRuleResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Configure the scoring rule for a tournament (admin only; creates a new version)",
)
async def upsert_scoring_rule(
    tournament_id: PydanticObjectId, payload: ScoringRuleUpsert
) -> DataResponse[ScoringRuleResponse]:
    rule = await result_service.upsert_scoring_rule(tournament_id, payload)
    return DataResponse(data=ScoringRuleResponse.model_validate(rule, from_attributes=True))


@admin_matches_router.post(
    "/{match_id}/results/manual",
    response_model=DataResponse[MatchResultResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Manually enter a raw match result by Game UID (admin only)",
)
async def enter_match_result(
    match_id: PydanticObjectId, payload: MatchResultCreate
) -> DataResponse[MatchResultResponse]:
    result = await result_service.enter_match_result(match_id, payload)
    return DataResponse(data=MatchResultResponse.model_validate(result, from_attributes=True))


@admin_results_router.put(
    "/{match_result_id}",
    response_model=DataResponse[MatchResultResponse],
    summary="Correct a raw match result before publication (admin only; recalculates the tournament result)",
)
async def update_match_result(
    match_result_id: PydanticObjectId, payload: MatchResultUpdate
) -> DataResponse[MatchResultResponse]:
    result = await result_service.update_match_result(match_result_id, payload)
    return DataResponse(data=MatchResultResponse.model_validate(result, from_attributes=True))


@admin_results_router.post(
    "/{match_result_id}/correction",
    response_model=DataResponse[CorrectionResponse],
    summary="Correct a result that may already be published (admin only; audited, recalculates ranks/scores)",
)
async def correct_result(
    match_result_id: PydanticObjectId,
    payload: MatchResultCorrection,
    current_user: User = Depends(get_current_user),
) -> DataResponse[CorrectionResponse]:
    match_result, revision = await result_service.correct_result(match_result_id, payload, current_user.id)
    return DataResponse(
        data=CorrectionResponse(
            match_result=MatchResultResponse.model_validate(match_result, from_attributes=True),
            revision=ResultRevisionResponse.model_validate(revision, from_attributes=True),
        )
    )


@admin_tournament_router.post(
    "/{tournament_id}/review-results",
    response_model=ListResponse[TournamentResultResponse],
    summary="Bulk-approve all draft calculated results for a tournament (admin only)",
)
async def review_results(tournament_id: PydanticObjectId) -> ListResponse[TournamentResultResponse]:
    results = await result_service.review_results(tournament_id)
    return ListResponse(data=[TournamentResultResponse.model_validate(r, from_attributes=True) for r in results])


@admin_tournament_router.post(
    "/{tournament_id}/publish-results",
    response_model=ListResponse[TournamentResultResponse],
    summary="Publish all approved results for a tournament (admin only); locks them against casual edits",
)
async def publish_results(tournament_id: PydanticObjectId) -> ListResponse[TournamentResultResponse]:
    results = await result_service.publish_results(tournament_id)
    return ListResponse(data=[TournamentResultResponse.model_validate(r, from_attributes=True) for r in results])

File: admin/src/api/admin/results.ts
import { http } from "@shared/api/client";
import {
  mapMatchResult,
  mapScoringRule,
  mapTournamentResult,
  type MatchResultDto,
  type ScoringRuleDto,
  type TournamentResultDto,
} from "@/lib/adminMappers";
import type { MatchResult, ResultRevision, ScoringMethod, ScoringRule, TournamentResult } from "@/types/admin";

// GET /api/tournaments/{id}/scoring-rule
export async function getScoringRule(tournamentId: string): Promise<ScoringRule> {
  const dto = await http.get<ScoringRuleDto>(`/api/tournaments/${tournamentId}/scoring-rule`);
  return mapScoringRule(dto);
}

export interface ScoringRuleInput {
  method: ScoringMethod;
  placementPoints?: Record<string, number>;
  killPointValue?: number;
  fieldWeights?: Record<string, number>;
  tieBreakFields?: string[];
}

// POST /api/admin/tournaments/{id}/scoring-rule (creates a new version)
export async function upsertScoringRule(tournamentId: string, input: ScoringRuleInput): Promise<ScoringRule> {
  const dto = await http.post<ScoringRuleDto>(`/api/admin/tournaments/${tournamentId}/scoring-rule`, {
    method: input.method,
    placement_points: input.placementPoints ?? null,
    kill_point_value: input.killPointValue ?? null,
    field_weights: input.fieldWeights ?? null,
    tie_break_fields: input.tieBreakFields ?? [],
  });
  return mapScoringRule(dto);
}

// GET /api/admin/matches/{id}/results
export async function listMatchResults(matchId: string): Promise<MatchResult[]> {
  const dtos = await http.get<MatchResultDto[]>(`/api/admin/matches/${matchId}/results`);
  return dtos.map(mapMatchResult);
}

// POST /api/admin/matches/{id}/results/manual
export async function enterMatchResult(
  matchId: string,
  input: { gameUid: string; placement?: number; kills?: number; evidenceReference?: string },
): Promise<MatchResult> {
  const dto = await http.post<MatchResultDto>(`/api/admin/matches/${matchId}/results/manual`, {
    game_uid: input.gameUid,
    placement: input.placement ?? null,
    kills: input.kills ?? null,
    evidence_reference: input.evidenceReference || undefined,
  });
  return mapMatchResult(dto);
}

// PUT /api/admin/results/{matchResultId} -- before publication
export async function updateMatchResult(
  matchResultId: string,
  input: { placement?: number; kills?: number },
): Promise<MatchResult> {
  const dto = await http.put<MatchResultDto>(`/api/admin/results/${matchResultId}`, {
    ...(input.placement !== undefined ? { placement: input.placement } : {}),
    ...(input.kills !== undefined ? { kills: input.kills } : {}),
  });
  return mapMatchResult(dto);
}

// POST /api/admin/results/{matchResultId}/correction -- may already be published; audited
export async function correctMatchResult(
  matchResultId: string,
  input: { placement?: number; kills?: number; reason: string },
): Promise<{ matchResult: MatchResult; revision: ResultRevision }> {
  const dto = await http.post<{
    match_result: MatchResultDto;
    revision: {
      id: string;
      tournament_id: string;
      match_result_id: string;
      tournament_result_id: string;
      user_id: string;
      changed_by: string;
      reason: string;
      changes: Record<string, { old: unknown; new: unknown }>;
      old_rank: number | null;
      new_rank: number | null;
      old_total_score: number | null;
      new_total_score: number | null;
      financial_impact_flagged: boolean;
      created_at: string;
    };
  }>(`/api/admin/results/${matchResultId}/correction`, {
    ...(input.placement !== undefined ? { placement: input.placement } : {}),
    ...(input.kills !== undefined ? { kills: input.kills } : {}),
    reason: input.reason,
  });
  return {
    matchResult: mapMatchResult(dto.match_result),
    revision: {
      id: String(dto.revision.id),
      tournamentId: String(dto.revision.tournament_id),
      matchResultId: String(dto.revision.match_result_id),
      tournamentResultId: String(dto.revision.tournament_result_id),
      userId: String(dto.revision.user_id),
      changedBy: String(dto.revision.changed_by),
      reason: dto.revision.reason,
      changes: dto.revision.changes,
      oldRank: dto.revision.old_rank,
      newRank: dto.revision.new_rank,
      oldTotalScore: dto.revision.old_total_score,
      newTotalScore: dto.revision.new_total_score,
      financialImpactFlagged: dto.revision.financial_impact_flagged,
      createdAt: dto.revision.created_at,
    },
  };
}

// GET /api/tournaments/{id}/results (published/calculated results, any status visible to admin)
export async function listTournamentResults(tournamentId: string): Promise<TournamentResult[]> {
  const dtos = await http.get<TournamentResultDto[]>(`/api/tournaments/${tournamentId}/results`);
  return dtos.map(mapTournamentResult);
}

// POST /api/admin/tournaments/{id}/review-results
export async function reviewResults(tournamentId: string): Promise<TournamentResult[]> {
  const dtos = await http.post<TournamentResultDto[]>(`/api/admin/tournaments/${tournamentId}/review-results`);
  return dtos.map(mapTournamentResult);
}

// POST /api/admin/tournaments/{id}/publish-results
export async function publishResults(tournamentId: string): Promise<TournamentResult[]> {
  const dtos = await http.post<TournamentResultDto[]>(`/api/admin/tournaments/${tournamentId}/publish-results`);
  return dtos.map(mapTournamentResult);
}
